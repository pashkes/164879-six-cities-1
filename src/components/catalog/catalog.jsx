import React from "react";
import PropTypes from "prop-types";
import Cities from "../cities/cities.jsx";
import Constants from "../../constants";
import Offers from "../offers/offers.jsx";
import Map from "../map/map.jsx";
import {
  getActiveCity,
  getFilteredOffers,
  getCoordinatesOfCurrentCity,
  getSelectedOffer,
} from "../../reducer/data/selectors";
import {connect} from "react-redux";

export const Catalog = (props) => {
  const {
    currentCity,
    offers,
    coordinates,
    currentOffer,
  } = props;
  return (
    <main className="page__main page__main--index">
      <h1 className="visually-hidden">Cities</h1>
      <div className="cities tabs">
        <Cities cities={Constants.CITIES}/>
      </div>
      <div className="cities__places-wrapper">
        <div className="cities__places-container container">
          <section className="cities__places places">
            <h2 className="visually-hidden">Places</h2>
            <b className="places__found">{offers.length} places to stay in {currentCity}</b>
            <form className="places__sorting" action="#" method="get">
              <span className="places__sorting-caption">Sort by</span>
              <span className="places__sorting-type" tabIndex="0">Popular<svg className="places__sorting-arrow" width="7" height="4"><use
                xlinkHref="#icon-arrow-select"/></svg></span>
              <ul className="places__options places__options--custom places__options--opened">
                <li className="places__option places__option--active" tabIndex="0">Popular</li>
                <li className="places__option" tabIndex="0">Price: low to high</li>
                <li className="places__option" tabIndex="0">Price: high to low</li>
                <li className="places__option" tabIndex="0">Top rated first</li>
              </ul>
              {/* <select className="places__sorting-type" id="places-sorting">
                <option className="places__option" value="popular" selected="">Popular</option>
                <option className="places__option" value="to-high">Price: low to high</option>
                <option className="places__option" value="to-low">Price: high to low</option>
                <option className="places__option" value="top-rated">Top rated first</option>
              </select>*/}
            </form>
            <Offers classModCard={`cities__place-card`} offers={offers} classModOffers={[`cities__places-list`, `tabs__content`]} />
          </section>
          <div className="cities__right-section">
            <section className="cities__map map">
              <Map selectedOffer={currentOffer} currentCity={currentCity} coordinates={coordinates} />
            </section>
          </div>
        </div>
      </div>
    </main>
  );
};

Catalog.propTypes = {
  currentCity: PropTypes.string.isRequired,
  offers: PropTypes.array.isRequired,
  coordinates: PropTypes.array.isRequired,
  currentOffer: PropTypes.array,
};

const mapStateToProps = (state) => ({
  currentCity: getActiveCity(state),
  offers: getFilteredOffers(state),
  coordinates: getCoordinatesOfCurrentCity(state),
  currentOffer: getSelectedOffer(state),
});

export {mapStateToProps};
export default connect(mapStateToProps)(Catalog);
