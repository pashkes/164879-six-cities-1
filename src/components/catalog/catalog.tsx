import * as React from "react";
import {connect} from "react-redux";
import {Offer} from "../../types";
import {SortOptions, TypeSort} from "../../constants";


import {
  getCurrentCity,
  getFilteredOffers,
  getSelectedOffer,
  getTypeSort,
  getCitiesFromServer,
} from "../../reducer/data/selectors";

import Map from "../map/map";
import Sorter from "./../select/select";
import Cities from "../cities/cities";
import Offers from "../offers/offers";
import MainEmpty from "./../main-empty/main-empty";

interface Props {
  currentCity: string,
  offers: Offer[],
  coordinates: [[number, number]],
  currentOffer: [number, number],
  cities: string[],
}

export const Catalog: React.FunctionComponent<Props> = (props) => {
  const {
    currentCity,
    offers,
    coordinates,
    currentOffer,
    cities,
  } = props;

  // @ts-ignore
  return (
    <React.Fragment>
      {
        offers.length !== 0 ?
          <main className="page__main page__main--index">
            <h1 className="visually-hidden">Cities</h1>
            <div className="cities tabs">
              <Cities cities={cities}/>
            </div>
            <div className="cities__places-wrapper">
              <div className="cities__places-container container">
                <section className="cities__places places">
                  <h2 className="visually-hidden">Places</h2>
                  <b className="places__found" role="status" aria-live="polite">{offers.length} places to stay in {currentCity}</b>
                  <form className="places__sorting" action="#" method="get">
                    <span className="places__sorting-caption">Sort by</span>
                    <Sorter options={SortOptions}/>
                  </form>
                  <Offers offers={offers} classModOffers={`cities__places-list tabs__content`}/>
                </section>
                <div className="cities__right-section">
                  <section className="cities__map map">
                    <Map selectedOffer={currentOffer} currentCity={currentCity} coordinates={coordinates}/>
                  </section>
                </div>
              </div>
            </div>
          </main>
          : <MainEmpty selectedCity={currentCity}/>
      }
    </React.Fragment>
  );
};

const sortOffers = (offers, sort) => {
  switch (sort.type) {
    case (TypeSort.TO_HIGH):
      return [...offers].sort((current, next) => current.price - next.price);
    case (TypeSort.TO_LOW):
      return [...offers].sort((current, next) => next.price - current.price);
    case (TypeSort.TOP_RATED):
      return [...offers].sort((current, next) => next.rating - current.rating);
    default:
      return offers;
  }
};

const mapStateToProps = (state) => {
  const offers = getFilteredOffers(state);
  const typeSort = getTypeSort(state);

  return {
    currentCity: getCurrentCity(state),
    offers: sortOffers(offers, typeSort),
    coordinates: offers.map((it) => [it.location.latitude, it.location.longitude]),
    currentOffer: getSelectedOffer(state),
    cities: getCitiesFromServer(state),
  };
};

export default connect(mapStateToProps)(Catalog);
