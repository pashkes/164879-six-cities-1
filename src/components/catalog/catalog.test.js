import React from 'react';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import {Catalog} from "./catalog.jsx";

configure({adapter: new Adapter()});

const inputOffers = [
  {
    id: ``,
    city: `Amsterdam`,
    title: ``,
    photoURL: ``,
    isPremium: false,
    isFavorite: false,
    price: 0,
    type: ``,
    rating: 0,
    coordinates: [0, 0]
  },
  {
    id: ``,
    city: `Berlin`,
    title: ``,
    photoURL: ``,
    isPremium: false,
    isFavorite: false,
    price: 0,
    type: ``,
    rating: 0,
    coordinates: [0, 0]
  }
];

it(`catalog renders correctly`, () => {
  const catalog = shallow(
      <Catalog
        offers={inputOffers}
        currentCity={`Amsterdam`}
      />
  );
  expect(catalog).toMatchSnapshot();
});
