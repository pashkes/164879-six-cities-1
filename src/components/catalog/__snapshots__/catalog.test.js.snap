// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalog renders correctly 1`] = `
<Fragment>
  <main
    className="page__main page__main--index"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="cities tabs"
    >
      <ConnectFunction
        cities={
          Array [
            "Paris",
            "Cologne",
            "Brussels",
            "Amsterdam",
            "Hamburg",
            "Dusseldorf",
          ]
        }
      />
    </div>
    <div
      className="cities__places-wrapper"
    >
      <div
        className="cities__places-container container"
      >
        <section
          className="cities__places places"
        >
          <h2
            className="visually-hidden"
          >
            Places
          </h2>
          <b
            aria-live="polite"
            className="places__found"
            role="status"
          >
            2
             places to stay in 
          </b>
          <form
            action="#"
            className="places__sorting"
            method="get"
          >
            <span
              className="places__sorting-caption"
            >
              Sort by
            </span>
            <WithSelect
              options={
                Array [
                  Object {
                    "name": "Popular",
                    "selected": true,
                    "value": "popular",
                  },
                  Object {
                    "name": "Price: low to high",
                    "selected": false,
                    "value": "to-high",
                  },
                  Object {
                    "name": "Price: high to low",
                    "selected": false,
                    "value": "to-low",
                  },
                  Object {
                    "name": "Top rated first",
                    "selected": false,
                    "value": "top-rated",
                  },
                ]
              }
            />
          </form>
          <ConnectFunction
            classModOffers={
              Array [
                "cities__places-list",
                "tabs__content",
              ]
            }
            offers={
              Array [
                Object {
                  "city": "Amsterdam",
                  "coordinates": Array [
                    0,
                    0,
                  ],
                  "id": 0,
                  "isFavorite": false,
                  "isPremium": false,
                  "previewPhoto": "",
                  "price": 0,
                  "rating": 0,
                  "title": "",
                  "type": "",
                },
                Object {
                  "city": "Berlin",
                  "coordinates": Array [
                    0,
                    0,
                  ],
                  "id": 1,
                  "isFavorite": false,
                  "isPremium": false,
                  "previewPhoto": "",
                  "price": 0,
                  "rating": 0,
                  "title": "",
                  "type": "",
                },
              ]
            }
          />
        </section>
        <div
          className="cities__right-section"
        >
          <section
            className="cities__map map"
          >
            <Map
              coordinates={
                Array [
                  0,
                  0,
                ]
              }
              currentCity=""
            />
          </section>
        </div>
      </div>
    </div>
  </main>
</Fragment>
`;
